### PR TITLE
Add support for DLT_LINUX_SLL2 capture format

### LINKED ISSUE #3591: [Enhancement] Add support for DLT_LINUX_SLL2
### Brief description

I have a capture file that uses [Linux "cooked" v2 encapsulation format](https://www.tcpdump.org/linktypes/LINKTYPE_LINUX_SLL2.html) which scapy cannot read currently. It would be nice to have support since the format is becoming more popular.

### Scapy version

2.4.5

### Python version

3.8

### Operating system

Linux 5.13.0

### Additional environment information

_No response_

### How to reproduce

Trying to read it gives following error 

```
$ cap = sniff(offline='/home/tsaarni/new.pcap')
WARNING: PcapReader: unknown LL type [276]/[0x114]. Using Raw packets
```

### Actual result

_No response_

### Expected result

_No response_

### Related resources

Currently tcpdump uses LINUX_SLL2 by default when [capturing "any" device](https://github.com/the-tcpdump-group/tcpdump/pull/850).
There is PR https://github.com/secdev/scapy/pull/3479 which intended to add support for the encapsulation, but it has not been updated for a while.

### PR DESCRIPTION
**Pull request by tsaarni**
**Checklist:**

-   [x] If you are new to Scapy: I have checked [CONTRIBUTING.md](https://github.com/secdev/scapy/blob/master/CONTRIBUTING.md) (esp. section submitting-pull-requests)
-   [x] I squashed commits belonging together
-   [x] I added unit tests or explained why they are not relevant
-   [x] I executed the regression tests for Python2 and Python3 (using `tox` or, `cd test && ./run_tests_py2, cd test && ./run_tests_py3`)
-   [x] If the PR is still not finished, please create a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

This PR adds support for CookedLinux v2 format ([DLT_LINUX_SLL2](https://www.tcpdump.org/linktypes/LINKTYPE_LINUX_SLL2.html)).

Fixes #3591
closes https://github.com/secdev/scapy/pull/3479

### CODE REVIEW

**Comment by tsaarni** [comment]
@Smartsmurf I see that you have previously created PR #3479. Since there has not been changes recently, I though I could try to fix this if you do not mind.

File: test/scapy/layers/l2.uts [hunk 1]
```diff
@@ -83,3 +83,27 @@ assert isinstance(p.payload, NoPayload)
...
+############
++ CookedLinux
+
+= CookedLinux - Basic Dissection
+
+cl = CookedLinux(b'\x00\x00\x03\x04\x00\x06\x00\x00\x00\x00\x00\x00\x6f\x50\x08\x00')
+assert cl.pkttype == 0 # unicast
+assert cl.lladdrtype == 772 # loopback
+assert cl.lladdrlen == 6
+assert cl.src == b'\x00\x00\x00\x00\x00\x00\x6f\x50'
```
Comment (tsaarni) [inline test/scapy/layers/l2.uts:98]: The `src` field is 8 bytes but address lenght may be e.g. only 6 bytes long. Only 6 bytes from the total 8 bytes should be meaningful ([LINKTYPE_LINUX_SLL](https://www.tcpdump.org/linktypes/LINKTYPE_LINUX_SLL.html) and [LINKTYPE_LINUX_SLL2](https://www.tcpdump.org/linktypes/LINKTYPE_LINUX_SLL2.html)).

Currently existing `CookedLinux` and newly proposed `CookedLinuxV2` use `StrFixedLenField("src", b"", 8)` which always includes all 8 bytes. For example, in this test case the `src` field contains 2 extra bytes that should not be there.  Should this be fixed?  If yes, are there examples of similar field that I could look for how to do this?

**Comment by Smartsmurf** [comment]
> @Smartsmurf I see that you have previously created PR #3479. Since there has not been changes recently, I though I could try to fix this if you do not mind.

Great. Thank you. Haven't got not much time in the last weeks to finalize.

File: test/scapy/layers/l2.uts (hunk 1 above)
Comment (gpotter2) [reply test/scapy/layers/l2.uts:98]: > Only 6 bytes from the total 8 bytes should be meaningful

This is the case with **some** linklayer addresses, but you could have any length by changing lladdrlen. So I'd say current implementation is fine

**Comment by othiman** [comment]
Hi. I stumbled across this because I used `tcpdump -i any` for capturing network traffic and at least recent versions of tcpdump use the CookedLinux v2 format ([DLT_LINUX_SLL2](https://www.tcpdump.org/linktypes/LINKTYPE_LINUX_SLL2.html)) doing so. I installed scapy with this patch by `pip install --user git+https://github.com/secdev/scapy.git@a04e4012bd45feeeade7b461fa687d48809510de` and afterwards sniff offline processing with my tcpdump files works just fine. Thanks for the nice patch! :-)
Best regards,
Thomas

File: scapy/layers/l2.py [hunk 2]
```diff
@@ -291,21 +292,38 @@ def l2_register_l3(l2, l3):
...
-                                                 3: "unicast-to-another-host",
-                                                 4: "sent-by-us"}),
+    fields_desc = [ShortEnumField("pkttype", 0, COOKED_LINUX_PACKET_TYPES),
                    XShortField("lladdrtype", 512),
                    ShortField("lladdrlen", 0),
                    StrFixedLenField("src", b"", 8),
                    XShortEnumField("proto", 0x800, ETHER_TYPES)]
 
 
+class CookedLinuxV2(Packet):
```
Comment (gpotter2) [inline scapy/layers/l2.py:315]: If you do the following:
```suggestion
class CookedLinuxV2(CookedLinux):
```
you then don't have do duplicate ALL the bindings 🙂

File: scapy/layers/l2.py [hunk 3]
```diff
@@ -652,6 +670,11 @@ class Dot1AD(Dot1Q):
 bind_layers(CookedLinux, Dot1AD, type=0x88a8)
 bind_layers(CookedLinux, Ether, proto=1)
 bind_layers(CookedLinux, ARP, proto=2054)
+bind_layers(CookedLinuxV2, LLC, proto=122)
+bind_layers(CookedLinuxV2, Dot1Q, proto=33024)
+bind_layers(CookedLinuxV2, Dot1AD, type=0x88a8)
+bind_layers(CookedLinuxV2, Ether, proto=1)
+bind_layers(CookedLinuxV2, ARP, proto=2054)
```
Comment (gpotter2) [inline scapy/layers/l2.py:677]: remove all of those

**Review by gpotter2 — CHANGES_REQUESTED** [review]
Hi ! Thanks a lot for the PR.
I have a comment: you don't need to duplicate all the bindings if you extend the class: this will be much cleaner.

File: scapy/layers/l2.py (hunk 2 above)
Comment (tsaarni) [reply scapy/layers/l2.py]: Ok, the upper layer bindings are done per class, so inherinting works here. Cool! :) 
Fix done.

File: scapy/layers/l2.py (hunk 3 above)
Comment (tsaarni) [reply scapy/layers/l2.py]: Done.

File: scapy/layers/eap.py [hunk 4]
```diff
@@ -770,6 +771,7 @@ def extract_padding(self, s):
 bind_layers(Ether, EAPOL, type=34958)
 bind_layers(Ether, EAPOL, dst='01:80:c2:00:00:03', type=34958)
 bind_layers(CookedLinux, EAPOL, proto=34958)
+bind_layers(CookedLinuxV2, EAPOL, proto=34958)
```
Comment (gpotter2) [inline scapy/layers/eap.py:774]: here..

**Review by gpotter2 — CHANGES_REQUESTED** [review]
There are a lot of places you still need to remove the changes from. Here are a few (incomplete)

File: scapy/layers/can.py [hunk 5]
```diff
@@ -144,6 +144,7 @@ def extract_padding(self, p):
 
 conf.l2types.register(DLT_CAN_SOCKETCAN, CAN)
 bind_layers(CookedLinux, CAN, proto=12)
+bind_layers(CookedLinuxV2, CAN, proto=12)
```
[the remark this answers is not in the document]
Comment (tsaarni) [reply scapy/layers/can.py]: Do'h of course, sorry for the confusion! It was some weeks ago that I did this so I already forgot and just reacted too quickly. I've now reverted all those changes.

**Review by gpotter2 — APPROVED** [review]
Thanks for the PR!

**Pull request merged**